### ISSUE TITLE
repo organization

### ISSUE THREAD

**Issue by kyleabeauchamp** [ParmEd/ParmEd#85]
I keep finding it counterintuitive that we have two top level package imports (`import chemistry, ParmedTools`).  Also hard to figure out which things are in which import.

Obviously this is probably a hard-to-change legacy issue, so feel free to close this issue.

**Comment by swails** [comment]
I have no problems reorganizing.  The API is stabilizing, but not fully stable yet, I don't think (I'm about ready to ensure backwards compatibility for all classes except those in `chemistry.tinker`, but I'm not set on package/module arrangement).

I separate these top-level packages in my head -- `chemistry` has all of the parsers, core classes, etc. while `ParmedTools` contains the interpreter machinery and GUI widgets.  The interpreter actions are really the only thing you'd want to import from there, but that works primarily on Amber prmtops (several have been generalized to work on `Structure`, but there's a definite Amber bias in `ParmedTools` that doesn't exist in `chemistry`).

No other package layout seemed superior to me, but as it's been like this for ca. 4 years now, I'm too familiar with it to be a good judge.  I'm open to any ideas/suggestions to improve the layout.

**Comment by kyleabeauchamp** [comment]
I guess my intial expectation was something like:

```
from parmed import chemistry, etc
```

But definitely ask around, as other users might have better thoughts.  Lee Ping and Robert also might be the people to ask.

**Comment by swails** [comment]
Hmm... that's an idea.  I'll have to figure out how to get that to play nicely with `parmed.py` -- it would be too easy to accidentally import the script the way things are laid out now (and neither `parmed.py` nor `xparmed.py` are import-friendly).

Maybe move `parmed.py` into `parmed/__main__.py` and dump a script into the bindir by `setup.py` somehow...

**Comment by swails** [comment]
I thought about this a little more and came up with a potential solution that would implement the behavior you're looking for while still maintaining backwards compatibility... but it doesn't feel all that elegant.

I could make `parmed.py` importable (with its execution protected behind an ``if __name__ == '__main__':`).  So it would look something like this:

```Python
#!/usr/bin/env python
import chemistry
import ParmedTools

if __name__ = '__main__':
    <script>
```

That way, `parmed.py` works both as the "top-level ParmEd namespace" as well as the ParmEd interpreter, but `chemistry` and `ParmedTools` still live in the global import namespace as well.  So then you could do stuff like

```Python
from parmed import chemistry
from parmed import ParmedTools
```

as well as what I usually do

```Python
from chemistry.amber import AmberParm
from chemistry.charmm import CharmmPsfFile
...
```

And then we'd pick one to document (that's the part I'm not really big on).  What do you think?

Note, this is similar to `os` importing and exposing `posixpath` as `path` (although not exactly the same, since `os` does that so it can import `ntpath` for windows and make it easier to support cross-platform).

**Comment by kyleabeauchamp** [comment]
I'm OK with anything here.  I suggest you postpone until other (potential) users have thought about this, as I wouldn't want you to make organizational changes without quorum.

**Comment by swails** [comment]
Do you think my proposal would result in something too confusing?  Moreso than the status quo or your suggestion?

User base isn't particularly large at present, and those that are using it don't have as much experience in the development process of large Python packages (like `mdtraj` or `msmbuilder`), although input from anyone that's watching this repo and has an opinion is welcome.

**Comment by kyleabeauchamp** [comment]
Your plan sounds reasonable to me

**Comment by radakb** [comment]
I've been fairly silent on development issues here, so I suppose I should chime in. I'm a bit of a stickler for having separate modules for unrelated things and also not overly ambitious names. I doubt it would be sensible or useful to tear everything apart at this point, but to me the parmed.py program/interface was always very separate from the underyling AmberParm class, etc. Having "chemistry" as its own package (module?) would make sense. As of now I mostly use the structure reading/writing stuff for editing and getting stuff like sequence information. Editing is not really on my list. Then again, I suppose that is not really a case to separate the structure class from things like AmberCrd, etc...

**Comment by swails** [comment]
This isn't really about the class hierarchy -- it's about the *package* hierarchy.

Should the `ParmedTools` and `chemistry` packages (which both define parts of the API used by `parmed.py`) be encapsulated in the same import-level namespace (so they are both subpackages in the same package). [1] Or should they be separate imports? [2]

[1] The "one program/library, one package" philosophy used by most (all?) libraries out there -- numpy, scipy, matplotlib, ipython, scikits-learn, pytables, ...

[2] `chemistry` has been globally importable since the original release of ParmEd, and contains the majority of functionality most people will want to use.

Still another option, similar to what `os.path` does, is to simply import `ParmedTools` into the `chemistry` namespace and make it accessible there (so that `chemistry` is the main package).  But it might seem like a project should provide a package whose name is at least loosely based on the name of the project :).

**Comment by radakb** [comment]
Right, I guess the subtext of that idea was that the package would also be split up. However that really makes no sense.

**Comment by hainm** [comment]
What's about

from mdtools import parmed 

from mdtools import chemistry

**Comment by swails** [comment]
`mdtools` is too generic.  So is `chemistry`, but there's not much I can do about that now, but fortunately no other packages that I know of use `chemistry` (PyMol uses `chempy`, and all other libraries I know of use the name of their library).

If I put everything in a top-level namespace, it should be named something like `parmed`...

**Comment by hainm** [comment]
ha ha,

I think `mdtools` is still much more specific than `chemistry`. 

    from parmed.chemistry import *
    from parmed import ParmedTools
    #from parmtools import chemisty
    #from parmtools import *

seems to be good candidate too.

btw, I myself prefer to have very short package's name like `numpy`, `scipy`, `pandas`, `mdtraj` (vs `mdanalysis` package) ... (thanks @swails  for suggesting to change `pycpptraj` to `pytraj`, it looks much nicer and saves lots of time). One of the reasons I still hesitate to try opnmm because it always has very long long methods' name and mixing with C/C++ style.

PS: In my opinion, ParmedTools is not a good name because:
* it's quite long
* the key combination is hard to type (at least for me)
* `parmed` reminds me about past tense verb rather than parm-edit

Hai

**Comment by swails** [comment]
> I think mdtools is still much more specific than chemistry.

Sure, but I already admitted `chemistry` is too general :), although I've been using it so long that I now like it.  The problem is that ParmEd does not try to be `mdtools` -- it conspicuously lacks functionality provided by packages like `pytraj` and `mdtraj`, and with the growing maturity of those tools I have no intention of jumping into that market (outside of *maybe* supporting reading some more of the trajectory formats).

> One of the reasons I still hesitate to try opnmm because it always has very long long methods' name and mixing with C/C++ style.

Long method and class names are not discouraged -- long package and module names are.  The latter recommendation is based on filesystem limitations (not all filesystems support arbitrarily long names, so the suggestion to use short, but descriptive, package and module names stems from the desire to promote portability, not save screen).  While I prefer the PEP-8 method naming scheme with_pot_holes better to the mixedCamelCase, I find the OMM class names (e.g., `CustomGBForce`, `HarmonicBondForce`, `NonbondedForce`, etc.) to be very good choices, and the mixedCamelCase class method style doesn't even reach a minor annoyance for me...  After all, PEP-8 also states: "... where an existing library has a different style, internal consistency is preferred."

> PS: In my opinion, ParmedTools is not a good name because:

I like this name.  It's not really that long (4 letters longer than common Python packages `sklearn` and `netCDF4`) and is descriptive -- it contains the tools used by the ParmEd interpreter.

> parmed reminds me about past tense verb rather than parm-edit

And several of the splash screen ASCII art pictures make reference to such :).

**Comment by swails** [comment]
I'm really starting to reconsider breaking backwards compatibility to remedy this.  I mentioned in #169 what I want to do, but I'll reiterate it here:

I want the `chemistry` package to be renamed to `parmed`, and then move `ParmedTools` into a `parmed.tools` subpackage.  This will break all imports in existing scripts, but the more I think about it, the more I think it's justified.  Right now, ParmEd is a little bit of a challenge to uninstall -- you need to delete `ParmedTools`, `chemistry`, and `fortranformat` (it used to be a little worse), and `ParmedTools` and `ParmedTools.ParmedActions` violate the naming conventions by going CamelCase.

It will be annoying for me to upgrade my scripts, but that should be easy with the following translations:

```Python
import chemistry
# to
import parmed as chemistry

import ParmedTools
# to
import parmed.tools as ParmedTools
```

And scripts that need updating will immediately give `ImportError`s and tracebacks that immediately tell you what needs to be changed.  If this will only significantly affect me, then I want to make the change.  But I can't keep putting off this decision.  More people are starting to use ParmEd, and I want to declare a stable API for the 2.0 release.

I'd like to put this to a vote if people care.

**Comment by kyleabeauchamp** [comment]
I like your new naming idea, but I am willing to go either way given the hassle on your end.

**Comment by swails** [comment]
I'm fine with the hassle.  I almost always do all imports at the top of my scripts, so it will be very easy for me to fix (and I know exactly the changes I have to make).

**Comment by leeping** [comment]
@swails : I don't know how to do this, but one option is to allow imports from both places, but throw a deprecation warning when importing from the old place.  That should give people enough time to update their scripts without crashing right away.  Then in the 2.1 release you can delete the old import.

**Comment by swails** [comment]
I thought about that, but I'm in the same boat as you -- I don't know how to do it, either.  I thought of two things:

1. Have `setup.py` automatically generate copies of `parmed` and `ParmedTools` in the install location.  Yuck.
2. Add a `chemistry` package that just has the following in its `__init__.py`:

```Python
import warnings as _warnings
warnings.warn('chemistry has been renamed to parmed and will be removed in the future',
              DeprecationWarning)
from parmed import *
```

That should work?  What's your vote on this?

**Comment by hainm** [comment]
good for me with this

```python
import warnings as _warnings
warnings.warn('chemistry has been renamed to parmed and will be removed in the future',
              DeprecationWarning)
from parmed import *
```

**Comment by swails** [comment]
@hainm -- you mean you tried it and it works?  Or that's an acceptable proposal?

**Comment by hainm** [comment]
@swails acceptable proposal. 

But I myself still don't like `parmed` name much, I just like past tense verb to me.

Hai

**Comment by kyleabeauchamp** [comment]
Because this:

```
                 ./       |      |        \.
               .:(        |i __ j|        ):`.
             .'   `._     |`::::'|     _.'    `.
           .'        "---.j `::' f.---"         `.
     _____/     ___    ____      __    __  ____   ___    
    |      \   |   |  |     `__'|  \  /  ||    | |   \    
    |  .-.  | .'   `| | .-.  |-/|   \/   || ___| |  . \   
    |  |_|  | |  i  | | |_| /"":|        || |    | | \ \  
    |       / | .^. | |    /::::|        || |__. | |  \ \ 
    |  ----'  | | | | |    \ :: |        ||  __| | |  |  )
    |  |     .' ''' `.|  |\ \   |  i  i  j| |    . | /  /  
    |  |     |   _   ||  | \ \  |  |\/|  || |__. | |.  / .
   [|  |     |  | |  ||  |  \ \ |  |  |  ||    | |    /   ].
  ] `--'     :--' `--::--'   \_|`--' ::--"|____|-"-- /    :[
  |      __  ::-'''`.:' "--.    .----::.----:: ,.---._    :|
  [  .-""  "`'              \  /      "      `'       `-. :].
 ]:.'                        \/                          `.:[
 |/                                                        \|

```

**Comment by hainm** [comment]
I see. ha ha.

**Comment by swails** [comment]
> But I myself still don't like parmed name much, I just like past tense verb to me.

Fair enough, but the name is here to stay :).

For the reason that @kyleabeauchamp gave and this:

```
                        Y O U      H A V E      B E E N
                               _
                            ==(W{==========-      /===-
                              ||  (.--.)         /===-_---~~~~~~~~~------____
                              | \_,|**|,__      |===-~___                _,-' `
                 -==\\        `\ ' `--'   ),    `//~\\   ~~~~`---.___.-~~
             ______-==|        /`\_. .__/\ \    | |  \\           _-~`
       __--~~~  ,-/-==\\      (   | .  |~~~~|   | |   `\        ,'
    _-~       /'    |  \\     )__/==0==-\<>/   / /      \      /
  .'        /       |   \\      /~\___/~~\/  /' /        \   /'
 /  ____  /         |    \`\.__/-~~   \  |_/'  /          \/'
/-'~    ~~~~~---__  |     ~-/~         ( )   /'        _--~`
                  \_|      /        _) | ;  ),   __--~~
                    '~~--_/      _-~/- |/ \   '-~ \             
                   {\__--_/}    / \\_>-|)<__\      \
                   /'   (_/  _-~  | |__>--<__|      | 
                  |   _/) )-~     | |__>--<__|      |   
                  / /~ ,_/       / /__>---<__/      |  
                 o-o _//        /-~_>---<__-~      / 
                 (^(~          /~_>---<__-      _-~              
                ,/|           /__>--<__/     _-~                 
             ,//('(          |__>--<__|     /                  .----_ 
            ( ( '))          |__>--<__|    |                 /' _---_~\
         `-)) )) (           |__>--<__|    |               /'  /     ~\`\
        ,/,'//( (             \__>--<__\    \            /'  //        ||
      ,( ( ((, ))              ~-__>--<_~-_  ~--____---~' _/'/        /'
    `~/  )` ) ,/|                 ~-_~>--<_/-__       __-~ _/ 
  ._-~//( )/ )) `                    ~~-'_/_/ /~~~~~~~__--~ 
   ;'( ')/ ,)(                              ~~~~~~~~~~ 
  ' ') '( (/
    '   '  `             
                                  P A R M E D
```

**Comment by davidlmobley** [comment]
Gotta love the ASCII art. Thought it was a lost art...

>I want the chemistry package to be renamed to parmed, and then move ParmedTools into a parmed.tools subpackage. This will break all imports in existing scripts, but the more I think about it, the more I think it's justified. Right now, ParmEd is a little bit of a challenge to uninstall -- you need to delete ParmedTools, chemistry, and fortranformat (it used to be a little worse), and ParmedTools and ParmedTools.ParmedActions violate the naming conventions by going CamelCase.

I totally agree with this and am absolutely in favor. I also think that *NOW* is the time (i.e. probably before your formal 2.0 release) since this is a "major version" type change. The `chemistry` name has bothered me since I started using it, in part because it seems overly broad/ambitious (beyond the scope of the project), and in part because most projects I use have module names that correspond with their GitHub names. As soon as I installed (without reading anything) I tried `import parmed` and of course nothing happened. It actually took me a few minutes to figure out what I was supposed to import (probably in part because the docs weren't obviously linked from the main GitHub page at the time). 

I am not the hugest fan of the parmed name (though it does good make for good ASCII puns) though ParmEd parses better in my head and makes me OK with it (there is a pronunciation difference for me), but I think it is good enough and the project is far enough along it doesn't merit changing the name.

**Comment by swails** [comment]
> I totally agree with this and am absolutely in favor. I also think that NOW is the time (i.e. probably before your formal 2.0 release) since this is a "major version" type change.

Fully agreed.  There's a little bit of time before the 2.0 release, since OpenMM has some nice new features since 6.2 that I want to use.  So I can't release ParmEd 2.0 until OpenMM 6.3 is released, which presumably gives me at least until after Telluride (June 14-18) since Peter will be there.

> I am not the hugest fan of the parmed name (though it does good make for good ASCII puns) though ParmEd parses better in my head and makes me OK with it (there is a pronunciation difference for me), but I think it is good enough and the project is far enough along it doesn't merit changing the name.

ParmEd is the real name, since it's a parameter-topology file editor (or at least that's how it started -- it does more but the emphasis is on easy parameter editing for rapid model prototyping).  I try to adhere to PEP 8 where possible (`ParmedTools.ParmedActions` came very early on before I knew about that), which is why I want to make the package name all-lowercase (and conda requires that, anyway for some reason).  I like the name because:

- The name reflected what it does (exclusively at first, though now it does more than just parmtop-editing, but that's still one of its foundations)
- It's short and easy to say as a word
- It made for funny (to me) ASCII splash images
- I just do :).

**Comment by leeping** [comment]
Hi Jason,

I support your planned package rename, if you're happy with it.

- Lee-Ping

**Comment by swails** [comment]
OK, thanks.  I'll work on this change after #170 is done.

I'm leaning against monkey-patching a deprecated `chemistry` package for a couple reasons:

- I'm exactly sure how the import mechanics would be changed by whatever approach I decide to take.
- There is some possible collision for people that have installed AmberTools 15 (where ParmEd gets installed to a location referenced by `PYTHONPATH`), since `PYTHONPATH` trumps standard install locations.  This will fortuitously avoid that collision, since the package names will be different.
- I don't think there are enough users yet to justify such an ugly band-aid.

But I'm open to counterarguments/suggestions (add it for the beta, then get rid of it, maybe?)

**Comment by leeping** [comment]
Hi Jason,

Another question I have (perhaps already addressed above) is that if your new version moves to a new package name, then users who update their ParmEd installation will get the new version in the new location - but their applications / scripts will still use the old version unless they go in and change their import statements.  A user might install the new version but get confused since their code is still loading the old version.  Perhaps your installer script can delete the files from old installations and/or tell users to update their import statements.

- Lee-Ping

**Comment by swails** [comment]
> Perhaps your installer script can delete the files from old installations and/or tell users to update their import statements.

Good catch.  This was one of my original reservations against this change that I had just forgotten about.  I still think it's worth taking the plunge now before it's really too late, though.

Simply printing a warning or message in setup.py won't cut it.  A fresh install generates so much output (it copies a *lot* of modules and compiles a couple extensions) that any warning message of that type would get lost.

I think deleting the files automatically is the best choice.  In general I don't like when setup scripts do that (e.g., I really despise OpenMM's `setup.py` script that automatically guts any previous installs), but this is a bit different, and only necessary in the transitional period.  I also don't think the added complexity of making this user-controllable behavior is warranted given the current situation.

Thanks for the heads-up.

**Comment by swails** [comment]
OK, so using `parmed` as the name of the importable package is not going to work.  The problem is that `parmed.py` cannot import anything called `parmed`, since it tries to import itself.  So while the library itself will work (and currently *does* in #173), the script will not.

Unfortunately, `parmed.py` *is* too established, with too many users in the Amber community, to change.  So the core package will need a different name.

My thoughts:

- `parmedlib`
- `libparmed`
- `parmed_lib`
- `pmd`

What do other people think?

**Comment by hainm** [comment]
one vote for `parmtools`, or `parmtoolz` (like `pytoolz`).

**Comment by hainm** [comment]
or `parmedit`?

**Comment by swails** [comment]
I think I prefer `parmedlib` to those suggestions.  Another idea: Rename `parmed.py` and `xparmed.py` to `parmed` and `xparmed`.  That way, they won't trigger as Python modules.

**Comment by hainm** [comment]
I think rename `parmed.py` to `parmed` is a good idea. Same as `cython`

`import cython` # used in module
`cython my_cython_file.pyx` # used in command line

**Comment by leeping** [comment]
Yes, I support removing the file extension from the executable scripts.

**Comment by rmcgibbo** [comment]
> OK, so using parmed as the name of the importable package is not going to work. The problem is that parmed.py cannot import anything called parmed, since it tries to import itself. So while the library itself will work (and currently does in #173), the script will not.

I think this can be worked around. You just have to remove the directory containing parmed.py from `sys.path`.

**Comment by rmcgibbo** [comment]
For example, I made a file called `numpy.py` that uses the numpy library. It contains the following code

```
import sys
import os
this_directory = os.path.dirname(os.path.abspath(__file__))
while this_directory in sys.path:
    sys.path.remove(this_directory)

import numpy as np
print(np.random.randn(10))
```

This script runs fine with both py2.7 and py3.4: `python path/to/script/numpy.py`.

**Comment by swails** [comment]
Thanks @rmcgibbo -- that did the trick.  The code chunk I actually used was slightly different (made sure I normalized symlinks and worked around case-insensitivity on most Mac filesystems):

```Python
scriptdir = os.path.realpath(os.path.abspath(os.path.split(__file__)[0]))
is_mac = 'darwin' in sys.platform.lower()
while True:
    for i, folder in enumerate(sys.path):
        folder = os.path.realpath(os.path.abspath(folder))
        if folder == scriptdir:
            break
        elif is_mac and folder.lower() == scriptdir.lower():
            break
    else:
        break
    sys.path.pop(i)
```
